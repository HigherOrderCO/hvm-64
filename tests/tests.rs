use std::{
  fs,
  path::{Path, PathBuf},
  str::FromStr,
};

use hvmc::{
  ast::{self, Net},
  host::Host,
  run::{self, Strict},
};
use insta::{assert_debug_snapshot, assert_display_snapshot, assert_snapshot};
use loaders::*;

mod loaders;

#[test]
fn test_era_era() {
  let net = parse_core("@main = * & * ~ *");
  let (rwts, net) = normal(net, 16);
  assert_snapshot!(Net::to_string(&net), @"*");
  assert_debug_snapshot!(rwts.total(), @"2");
}

#[test]
fn test_era_era2() {
  let net = parse_core("@main = (* *) & * ~ *");
  let (rwts, net) = normal(net, 16);
  assert_snapshot!(Net::to_string(&net), @"(* *)");
  assert_debug_snapshot!(rwts.total(), @"2");
}

#[test]
fn test_commutation() {
  let net = parse_core("@main = root & (x x) ~ [* root]");
  let (rwts, net) = normal(net, 16);
  assert_snapshot!(Net::to_string(&net), @"(a a)");
  assert_debug_snapshot!(rwts.total(), @"5");
}

#[test]
fn test_bool_and() {
  let book = parse_core(
    "
    @true = (b (* b))
    @fals = (* (b b))
    @and  = ((b (@fals c)) (b c))
    @main = root & @and ~ (@true (@fals root))
  ",
  );
  let (rwts, net) = normal(book, 64);

  assert_snapshot!(Net::to_string(&net), @"(* (a a))");
  assert_debug_snapshot!(rwts.total(), @"9");
}

#[derive(Default)]
pub struct TestOpts {
  _lazy: bool,
}

fn test_host(name: &str, host: Host, _opts: TestOpts) {
  println!("{name}");
  let Some(entrypoint) = host.defs.get("main") else { return };
  let heap = run::Net::<Strict>::init_heap(1 << 28);
  let mut net = run::Net::<Strict>::new(&heap);
  net.boot(entrypoint);
  net.normal();

  assert_display_snapshot!(format!("{name}/strict"), host.readback(&net));
  assert_display_snapshot!(format!("{name}/strict_rwts"), net.rwts.total());

  // Now, test lazy mode and
  // ensure the outputs are equal.
}

fn test_path(path: &Path, opts: TestOpts) {
  println!("Exec {path:?}");
  let code = fs::read_to_string(&path).unwrap();
  let book = ast::Book::from_str(&code).unwrap();
  let host = Host::new(&book);

  let path = path.strip_prefix(env!("CARGO_MANIFEST_DIR")).unwrap();

  test_host(path.to_str().unwrap(), host, opts);
}

fn test_dir(dir: &Path, filter: &dyn Fn(&Path) -> bool) {
  insta::glob!(dir, "**/*.hvmc", |p| {
    if filter(p) {
      test_path(p, Default::default());
    }
  })
}

fn is_slow(p: &Path) -> bool {
  p.components().any(|x| x.as_os_str().to_str().unwrap() == "slow")
}

fn manifest_relative(sub: &str) -> PathBuf {
  format!("{}/{}", env!("CARGO_MANIFEST_DIR"), sub).into()
}

#[test]
fn test_fast_programs() {
  test_dir(&manifest_relative("tests/programs/"), &|p| !is_slow(p))
}

#[test]
#[ignore = "very slow"]
fn test_slow_programs() {
  test_dir(&manifest_relative("tests/programs/"), &|p| is_slow(p))
}

#[test]
fn test_fast_examples() {
  test_dir(&manifest_relative("examples/"), &|p| !is_slow(p));
}

#[test]
#[ignore = "very slow"]
fn test_slow_examples() {
  test_dir(&manifest_relative("examples/"), &|p| is_slow(p))
}
