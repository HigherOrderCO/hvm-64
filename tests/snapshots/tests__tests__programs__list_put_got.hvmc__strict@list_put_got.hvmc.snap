---
source: tests/tests.rs
expression: host.readback(&net)
input_file: tests/programs/list_put_got.hvmc
---
(((* (((* (((* (a a)) b) b)) ((* (((* ((((@S1$S0 c) (* c)) d) d)) ((* (((* ((((@S2$S1 e) (* e)) f) f)) ((* (((* ((((@S3$S2 g) (* g)) h) h)) ((* (((* ((((@S4$S3 i) (* i)) j) j)) ((* (((* ((((@S5$S4 k) (* k)) l) l)) ((* (((* ((((@S6$S5 m) (* m)) n) n)) ((* (((* ((((@S7$S6 o) (* o)) p) p)) ((* (((* ((((@S8$S7 q) (* q)) r) r)) ((* (((* ((((@S9$S8 s) (* s)) t) t)) ((* (((* ((((@S10$S9 u) (* u)) v) v)) ((* (((* ((((@S11$S10 w) (* w)) x) x)) ((* (((* ((((@S12$S11 y) (* y)) z) z)) ((* (((* ((((@S13$S12 aa) (* aa)) ab) ab)) ((* (((* ((((@S14$S13 ac) (* ac)) ad) ad)) ((* (((* ((((@S15$S14 ae) (* ae)) af) af)) ((* (((* ((((@S16$S15 ag) (* ag)) ah) ah)) ((* (((* ((((@S17$S16 ai) (* ai)) aj) aj)) ((* (((* ((((@S18$S17 ak) (* ak)) al) al)) ((* (((* ((((@S19$S18 am) (* am)) an) an)) ((* (((* ((((@S20$S19 ao) (* ao)) ap) ap)) ((* (((* ((((@S21$S20 aq) (* aq)) ar) ar)) ((* (((* ((((@S22$S21 as) (* as)) at) at)) ((* (((* ((((@S23$S22 au) (* au)) av) av)) ((* (((* ((((@S24$S23 aw) (* aw)) ax) ax)) ((* (((* ((((@S25$S24 ay) (* ay)) az) az)) ((* (((* ((((@S26$S25 ba) (* ba)) bb) bb)) ((* (((* ((((@S27$S26 bc) (* bc)) bd) bd)) ((* (((* ((((@S28$S27 be) (* be)) bf) bf)) ((* (((* ((((@S29$S28 bg) (* bg)) bh) bh)) ((* (((* ((((@S30$S29 bi) (* bi)) bj) bj)) ((* (((* ((((@S31$S30 bk) (* bk)) bl) bl)) ((bm (* bm)) bn)) bn)) bo)) bo)) bp)) bp)) bq)) bq)) br)) br)) bs)) bs)) bt)) bt)) bu)) bu)) bv)) bv)) bw)) bw)) bx)) bx)) by)) by)) bz)) bz)) ca)) ca)) cb)) cb)) cc)) cc)) cd)) cd)) ce)) ce)) cf)) cf)) cg)) cg)) ch)) ch)) ci)) ci)) cj)) cj)) ck)) ck)) cl)) cl)) cm)) cm)) cn)) cn)) co)) co)) cp)) cp)) cq)) cq)) cr)) cr)) cs)) cs)) [ct *]) ct)
